pkg/tbtc/heartbeat.go

```diff
@@ -111,16 +138,72 @@ func (ha *heartbeatAction) execute() error {
 	)
 	defer cancelHeartbeatCtx()
 
-	signature, _, err := ha.signingExecutor.sign(heartbeatCtx, messageToSign, ha.startBlock)
+	signature, activeOperatorsCount, signingEndBlock, err := ha.signingExecutor.sign(
+		heartbeatCtx,
+		messageToSign,
+		ha.startBlock,
+	)
+
+	// If there was no error and the number of active operators during signing
+	// was enough, we can consider the heartbeat procedure as successful.
+	if err == nil && activeOperatorsCount >= heartbeatSigningMinimumActiveOperators {
+		logger.Infof(
+			"successfully generated signature [%s] for heartbeat message [0x%x]",
+			signature,
+			ha.proposal.Message[:],
+		)
+
+		// Reset the counter for consecutive heartbeat failure.
+		*ha.failureCounter = 0
+
+		return nil
+	}
+
+	// If there was an error or the number of active operators during signing
+	// was not enough, we must consider the heartbeat procedure as a failure.
 	if err != nil {
-		return fmt.Errorf("cannot sign heartbeat message: [%v]", err)
+		logger.Infof("error while generating heartbeat signature: [%v]", err)
+	} else {
+		logger.Infof(
+			"not enough active operators during signing; required [%d]: "+
+				"actual [%d]",
+			activeOperatorsCount,
+			heartbeatSigningMinimumActiveOperators,
+		)
 	}
 
-	logger.Infof(
-		"generated signature [%s] for heartbeat message [0x%x]",
-		signature,
-		ha.proposal.Message[:],
+	// Increment the heartbeat failure counter.
+	*ha.failureCounter++
+
+	// If the number of consecutive heartbeat failures does not exceed the
+	// threshold do not notify about operator inactivity.
+	if *ha.failureCounter < heartbeatConsecutiveFailureThreshold {
+		logger.Infof(
+			"leaving without notifying about operator inactivity; current "+
+				"heartbeat failure count is [%d]",
+			*ha.failureCounter,
+		)
+		return nil
+	}
+
+	// The value of consecutive heartbeat failures exceeds the threshold.
+	// Proceed with operator inactivity notification.
+	err = ha.inactivityClaimExecutor.publishClaim(
+		// Leave the list of inactive operators empty even if some operators
+		// were inactive during signing heartbeat. The inactive operators could
+		// simply be in the process of unstaking and therefore should not be
+		// punished.
+		[]group.MemberIndex{},
+		true,
+		messageToSign,
+		signingEndBlock,
```

Comment:
If signing returns an error, `signingEndBlock` will be `0`. I think `publishClaim` does not handle this case.

Reply:
Good catch!
If we cannot rely on `signingEndBlock` returned from signing then perhaps we could use `startBlock` or `expiryBlock` stored into the `heartbeatAction`?

Couldn't `expiryBlock` be used as the timeout block for inactivity submission?

Reply:
I think the easiest way is to make `inactivityClaimExecutor.publishClaim` context-driven. 

Then, in `heartbeatAction.execute` we need to split `heatbeatCtx` into two:
- `heartbeatSigningCtx` that will steer the signing executor
- `heartbeatClaimCtx` that will control the inactivity claim executor

The `heartbeatSigningCtx` should expire in the middle of `heartbeatProposalValidityBlocks`. The `heartbeatClaimCtx` at the end. Both should maintain a safety margin of ~5 minutes (25 blocks). In that configuration we may need to extend `heartbeatProposalValidityBlocks` to e.g. 600 blocks (see https://github.com/keep-network/keep-core/pull/3800#discussion_r1590995784) but leaving that decision to you as I'm not sure what is the execution time of the inactivity claim.

Reply:
Done in 9ad6371c9f14856f97900d95e627f05ed29dc344.